File: Sources/PaperCore/PaperCandidateMerger.swift
import Foundation

public struct PaperCandidateMerger {
    public init() {}

    public func merge(_ candidates: [PaperCandidate]) -> [PaperCandidate] {
        var groups: [[PaperCandidate]] = []

        for candidate in candidates {
            let matchingIndexes = groups.indices.filter { index in
                groups[index].contains { candidatesMatch($0, candidate) }
            }

            guard let firstIndex = matchingIndexes.first else {
                groups.append([candidate])
                continue
            }

            groups[firstIndex].append(candidate)
            for index in matchingIndexes.dropFirst().reversed() {
                groups[firstIndex] += groups[index]
                groups.remove(at: index)
            }
        }

        return groups.map(mergeGroup).sorted(by: candidateComesFirst)
    }

    private func mergeGroup(_ group: [PaperCandidate]) -> PaperCandidate {
        let ordered = group.sorted(by: candidateComesFirst)
        var merged = ordered[0]
        merged.baseID = firstNonempty(ordered.map(\.baseID))
        merged.doi = firstNonempty(ordered.map(\.doi))
        merged.title = firstNonempty(ordered.map(\.title)) ?? merged.title
        merged.summary = firstNonempty(ordered.map(\.summary)) ?? ""
        merged.authors = orderedUnion(ordered.map(\.authors))
        merged.institutions = orderedUnion(ordered.map(\.institutions))
        merged.categories = orderedUnion(ordered.map(\.categories))
        merged.publishedAt = ordered.compactMap(\.publishedAt).min()
        merged.updatedAt = ordered.compactMap(\.updatedAt).max()
        merged.absURL = firstURL(ordered.map(\.absURL))
        merged.pdfURL = firstURL(ordered.map(\.pdfURL))
        merged.venue = firstNonempty(ordered.map(\.venue))
        merged.citationCount = ordered.compactMap(\.citationCount).max()
        merged.provenance = orderedUnion(ordered.map { candidate in
            candidate.provenance.isEmpty
                ? [PaperProvenance(source: candidate.source, sourceID: candidate.sourceID, sourceURL: candidate.absURL)]
                : candidate.provenance
        })
        merged.openAccessEvidence = preferredEvidence(from: ordered)
        merged.openAccessPDFURL = firstURL(ordered.map(\.openAccessPDFURL)) ?? merged.openAccessEvidence?.url
        return merged
    }

    private func candidatesMatch(_ lhs: PaperCandidate, _ rhs: PaperCandidate) -> Bool {
        let lhsDOI = normalizedDOI(lhs.doi)
        let rhsDOI = normalizedDOI(rhs.doi)
        if let lhsDOI, let rhsDOI {
            return lhsDOI == rhsDOI
        }

        let lhsArxivID = normalizedArxivID(lhs)
        let rhsArxivID = normalizedArxivID(rhs)
        if let lhsArxivID, let rhsArxivID {
            return lhsArxivID == rhsArxivID
        }

        guard let lhsTitleHash = normalizedTitleHash(lhs.title),
              let rhsTitleHash = normalizedTitleHash(rhs.title) else {
            return false
        }
        return lhsTitleHash == rhsTitleHash
    }

    private func normalizedDOI(_ doi: String?) -> String? {
        guard var value = doi?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(), !value.isEmpty else {
            return nil
        }
        for prefix in ["https://doi.org/", "http://doi.org/", "doi:"] where value.hasPrefix(prefix) {
            value.removeFirst(prefix.count)
        }
        return value.trimmingCharacters(in: CharacterSet(charactersIn: "."))
    }

    private func normalizedArxivID(_ candidate: PaperCandidate) -> String? {
        let value: String?
        if let baseID = candidate.baseID, !baseID.isEmpty {
            value = baseID
        } else if candidate.source == .arxiv {
            value = candidate.sourceID
        } else {
            value = nil
        }
        return value.map(PaperCandidate.arxivBaseID)?.lowercased()
    }

    private func normalizedTitleHash(_ title: String) -> String? {
        let normalized = title.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Locale(identifier: "en_US_POSIX"))
            .unicodeScalars
            .filter { CharacterSet.alphanumerics.contains($0) }
            .map(String.init)
            .joined()
        guard !normalized.isEmpty else { return nil }

        var hash: UInt64 = 1_469_598_103_934_665_603
        for byte in normalized.utf8 {
            hash ^= UInt64(byte)
            hash &*= 1_099_511_628_211
        }
        return String(format: "%016llx", hash)
    }

    private func candidateComesFirst(_ lhs: PaperCandidate, _ rhs: PaperCandidate) -> Bool {
        let lhsPriority = sourcePriority(lhs.source)
        let rhsPriority = sourcePriority(rhs.source)
        if lhsPriority != rhsPriority { return lhsPriority < rhsPriority }
        if lhs.sourceID.caseInsensitiveCompare(rhs.sourceID) != .orderedSame {
            return lhs.sourceID.caseInsensitiveCompare(rhs.sourceID) == .orderedAscending
        }
        return lhs.title.caseInsensitiveCompare(rhs.title) == .orderedAscending
    }

    private func sourcePriority(_ source: PaperSourceKind) -> Int {
        switch source {
        case .arxiv: 0
        case .semanticScholar: 1
        case .openAlex: 2
        case .crossref: 3
        case .unpaywall: 4
        case .web: 5
        }
    }

    private func firstNonempty(_ values: [String?]) -> String? {
        values.compactMap { value -> String? in
            guard let value else { return nil }
            let cleaned = value.cleanedWhitespace
            return cleaned.isEmpty ? nil : cleaned
        }.first
    }

    private func firstURL(_ values: [URL?]) -> URL? {
        values.compactMap { $0 }.first
    }

    private func orderedUnion<T: Hashable>(_ collections: [[T]]) -> [T] {
        var seen = Set<T>()
        return collections.flatMap { $0 }.filter { seen.insert($0).inserted }
    }

    private func preferredEvidence(from candidates: [PaperCandidate]) -> OpenAccessEvidence? {
        candidates.compactMap(\.openAccessEvidence).sorted { lhs, rhs in
            let lhsVerified = lhs.status == .verified
            let rhsVerified = rhs.status == .verified
            if lhsVerified != rhsVerified { return lhsVerified }
            if lhs.source != rhs.source { return sourcePriority(lhs.source) < sourcePriority(rhs.source) }
            return (lhs.url?.absoluteString ?? "") < (rhs.url?.absoluteString ?? "")
        }.first
    }
}

File: Apps/PaperPulseWindows/src/PaperPulse.Engine/PaperCandidateMerger.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using PaperPulse.Contracts;

namespace PaperPulse.Engine;

public sealed class PaperCandidateMerger
{
    public IReadOnlyList<PaperCandidate> Merge(IEnumerable<PaperCandidate> candidates)
    {
        List<List<PaperCandidate>> groups = [];
        foreach (PaperCandidate candidate in candidates)
        {
            List<int> matches = groups.Select((group, index) => (group, index))
                .Where(pair => pair.group.Any(existing => CandidatesMatch(existing, candidate)))
                .Select(pair => pair.index)
                .ToList();
            if (matches.Count == 0)
            {
                groups.Add([candidate]);
                continue;
            }

            int first = matches[0];
            groups[first].Add(candidate);
            foreach (int index in matches.Skip(1).OrderByDescending(index => index))
            {
                groups[first].AddRange(groups[index]);
                groups.RemoveAt(index);
            }
        }

        return groups.Select(MergeGroup).OrderBy(candidate => SourcePriority(candidate.Source))
            .ThenBy(candidate => candidate.SourceId, StringComparer.OrdinalIgnoreCase)
            .ThenBy(candidate => candidate.Title, StringComparer.OrdinalIgnoreCase)
            .ToList();
    }

    private static PaperCandidate MergeGroup(IEnumerable<PaperCandidate> group)
    {
        List<PaperCandidate> ordered = group.OrderBy(candidate => SourcePriority(candidate.Source))
            .ThenBy(candidate => candidate.SourceId, StringComparer.OrdinalIgnoreCase)
            .ThenBy(candidate => candidate.Title, StringComparer.OrdinalIgnoreCase)
            .ToList();
        PaperCandidate primary = ordered[0];
        List<int> citationCounts = ordered.Where(candidate => candidate.CitationCount.HasValue)
            .Select(candidate => candidate.CitationCount!.Value)
            .ToList();
        List<DateTimeOffset> publishedDates = ordered.Where(candidate => candidate.PublishedAt.HasValue)
            .Select(candidate => candidate.PublishedAt!.Value)
            .ToList();
        List<DateTimeOffset> updatedDates = ordered.Where(candidate => candidate.UpdatedAt.HasValue)
            .Select(candidate => candidate.UpdatedAt!.Value)
            .ToList();
        OpenAccessEvidence? evidence = ordered.Select(candidate => candidate.OpenAccessEvidence)
            .OfType<OpenAccessEvidence>()
            .OrderByDescending(item => item.Status == OpenAccessStatus.Verified)
            .ThenBy(item => SourcePriority(item.Source))
            .ThenBy(item => item.Url?.AbsoluteUri, StringComparer.Ordinal)
            .FirstOrDefault();

        return new PaperCandidate(
            primary.Source,
            primary.SourceId,
            FirstNonempty(ordered.Select(candidate => candidate.Title)) ?? primary.Title,
            FirstNonempty(ordered.Select(candidate => candidate.Summary)) ?? string.Empty,
            baseId: FirstNonempty(ordered.Select(candidate => candidate.BaseId)),
            doi: FirstNonempty(ordered.Select(candidate => candidate.Doi)),
            authors: OrderedUnion(ordered.Select(candidate => candidate.Authors)),
            institutions: OrderedUnion(ordered.Select(candidate => candidate.Institutions)),
            categories: OrderedUnion(ordered.Select(candidate => candidate.Categories)),
            publishedAt: publishedDates.Count == 0 ? null : publishedDates.Min(),
            updatedAt: updatedDates.Count == 0 ? null : updatedDates.Max(),
            absUrl: ordered.Select(candidate => candidate.AbsUrl).FirstOrDefault(url => url is not null),
            pdfUrl: ordered.Select(candidate => candidate.PdfUrl).FirstOrDefault(url => url is not null),
            venue: FirstNonempty(ordered.Select(candidate => candidate.Venue)),
            citationCount: citationCounts.Count == 0 ? null : citationCounts.Max(),
            openAccessPdfUrl: ordered.Select(candidate => candidate.OpenAccessPdfUrl).FirstOrDefault(url => url is not null) ?? evidence?.Url,
            provenance: OrderedUnion(ordered.Select(candidate => candidate.Provenance.Count == 0
                ? (IEnumerable<PaperProvenance>)new[] { new PaperProvenance { Source = candidate.Source, SourceId = candidate.SourceId, SourceUrl = candidate.AbsUrl } }
                : candidate.Provenance)),
            openAccessEvidence: evidence);
    }

    private static bool CandidatesMatch(PaperCandidate left, PaperCandidate right)
    {
        string? leftDoi = NormalizeDoi(left.Doi);
        string? rightDoi = NormalizeDoi(right.Doi);
        if (leftDoi is not null && rightDoi is not null) return leftDoi == rightDoi;

        string? leftArxiv = NormalizedArxivId(left);
        string? rightArxiv = NormalizedArxivId(right);
        if (leftArxiv is not null && rightArxiv is not null) return leftArxiv == rightArxiv;

        string? leftTitle = NormalizedTitleHash(left.Title);
        string? rightTitle = NormalizedTitleHash(right.Title);
        return leftTitle is not null && leftTitle == rightTitle;
    }

    public static string? NormalizeDoi(string? doi)
    {
        if (string.IsNullOrWhiteSpace(doi)) return null;
        string normalized = doi.Trim().ToLowerInvariant();
        foreach (string prefix in new[] { "https://doi.org/", "http://doi.org/", "doi:" })
        {
            if (normalized.StartsWith(prefix, StringComparison.Ordinal)) normalized = normalized[prefix.Length..];
        }
        normalized = normalized.TrimEnd('.');
        return normalized.Length == 0 ? null : normalized;
    }

    public static string? NormalizedArxivId(PaperCandidate candidate)
    {
        string? value = !string.IsNullOrWhiteSpace(candidate.BaseId)
            ? candidate.BaseId
            : candidate.Source == PaperSourceKind.Arxiv ? candidate.SourceId : null;
        return string.IsNullOrWhiteSpace(value) ? null : ArxivAtomParser.ArxivBaseId(value).ToLowerInvariant();
    }

    public static string? NormalizedTitleHash(string title)
    {
        string normalized = new string(title.Normalize(NormalizationForm.FormD)
            .Where(character => char.IsLetterOrDigit(character))
            .Select(char.ToLowerInvariant)
            .ToArray());
        if (normalized.Length == 0) return null;

        ulong hash = 1_469_598_103_934_665_603;
        foreach (byte value in Encoding.UTF8.GetBytes(normalized))
        {
            hash ^= value;
            hash *= 1_099_511_628_211;
        }
        return hash.ToString("x16", CultureInfo.InvariantCulture);
    }

    public static int SourcePriority(PaperSourceKind source) => source switch
    {
        PaperSourceKind.Arxiv => 0,
        PaperSourceKind.SemanticScholar => 1,
        PaperSourceKind.OpenAlex => 2,
        PaperSourceKind.Crossref => 3,
        PaperSourceKind.Unpaywall => 4,
        PaperSourceKind.Web => 5,
        _ => int.MaxValue
    };

    private static string? FirstNonempty(IEnumerable<string?> values) => values.Select(value => value?.Trim())
        .FirstOrDefault(value => !string.IsNullOrWhiteSpace(value));

    private static List<T> OrderedUnion<T>(IEnumerable<IEnumerable<T>> collections) where T : notnull
    {
        HashSet<T> seen = [];
        return collections.SelectMany(collection => collection).Where(seen.Add).ToList();
    }
}
